### PR TITLE
enhancement(vrl): Improved performance of starts_with

### PR DESCRIPTION
**Pull request by StephenWakely**
Ref #6680

This improves the performance of starts with.

If case sensitive we just compare the bytes without any utf8 checking, if insensitive we create an iterator so we only need to perform the utf8 checks on the part of the text we are actually interested in.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

**Review by jszwedko — APPROVED** [review]
👍 implementation looks good. I'm curious to see what the benchmarks report.

File: lib/vrl/stdlib/Cargo.toml [hunk 1]
```diff
@@ -42,6 +42,9 @@ uuid = { version = "0.8", features = ["v4"], optional = true }
 roxmltree = { version = "0.14.1", optional = true }
 woothee = { version = "0.13.0", optional = true }
 uaparser = { version = "0.4.0", optional = true }
+cached = { version = "0.26.2", optional = true }
```
Comment (blt) [inline lib/vrl/stdlib/Cargo.toml:45]: Is this used in the PR?

**Comment by blt** [comment]
I find the `http_to_http_acks` ding to be surprising here. It _could_ be that acking is erratic enough that we're seeing two distinct sequences from baseline and comparison naturally, unrelated to your change. I've re-triggered the soaks. 

REF #10290

**Comment by leebenson** [comment]
Is there a soak that already makes heavy use of starts_with?

It looks like numbers are generally improved all round against baselines, but I can't tell which might be specifically benefiting from this change.

**Comment by blt** [comment]
> Is there a soak that already makes heavy use of starts_with?
> 
> It looks like numbers are generally improved all round against baselines, but I can't tell which might be specifically benefiting from this change.

Only `http_pipelines_blackhole`, looks like. 

The difference in means is 3.29% with a 25%-ish confidence of that result being different from random chance. That jives with the last flamegraphs I saw, but I admit to being 4/5 days out of touch there. If the micro numbers look improved this'd be consistent with an optimization that didn't get at whatever bottlenecks `http_pipelines_blackhole` in practice for the SHAs experimented with.

**Comment by blt** [comment]
> Micro benches look good:
> 
> ```
> 2021-12-06T20:29:43.4493004Z vrl_stdlib/functions/starts_with/case_sensitive                                      97.518 ns  -33.135%     10.255 Melem/s  +49.556%           0.00  improved
> 2021-12-06T20:29:43.4493988Z vrl_stdlib/functions/starts_with/case_insensitive                                    117.08 ns  -39.756%     8.5410 Melem/s  +65.991%           0.00  improved
> ```

Agreed. Quality result. 

> That is curious that it had a negative effect on `http_to_http_acks` though, which appears to be consistent?

Yeah, very strange. I would like to get some understanding there before we merge up. Curious if @bruceg has any thoughts? If `starts_with` is used in either of the HTTP sink/source it's feasible the optimization here has moved a bottleneck to a worse place in that configuration. 

> `http_pipelines_blackhole` makes heavy use of `starts_with` but seemingly that is not a bottleneck yet.

Yep, at least not for that vector configuration, load generation. Still, not moving backward so that's a positive.

**Comment by jszwedko** [comment]
Yeah, what's weird is `remap` isn't even used in that soak: https://github.com/vectordotdev/vector/blob/master/soaks/tests/http_to_http_acks/terraform/vector.toml

**Comment by blt** [comment]
Right, yeah. Considering the consistency there must be a dependency on the changed code in either the source or sink there, even though the remap transform itself is not used directly.

**Comment by bruceg** [comment]
> If `starts_with` is used in either of the HTTP sink/source it's feasible the optimization here has moved a bottleneck to a worse place in that configuration.

I am not aware of `starts_with` being used anywhere in that stack. I would chalk this up to a fluke (instruction cache alignment, etc), except that it's larger than I would expect from that.

**Comment by blt** [comment]
> > If `starts_with` is used in either of the HTTP sink/source it's feasible the optimization here has moved a bottleneck to a worse place in that configuration.
> 
> I am not aware of `starts_with` being used anywhere in that stack. I would chalk this up to a fluke (instruction cache alignment, etc), except that it's larger than I would expect from that.

I will also note that this result is present [here](https://github.com/vectordotdev/vector/pull/10265#issuecomment-987330706) in Luke's work.

File: lib/vrl/stdlib/Cargo.toml (hunk 1 above)
Comment (StephenWakely) [reply lib/vrl/stdlib/Cargo.toml:45]: No, I'm not sure how that got there.

File: lib/vrl/stdlib/Cargo.toml [hunk 2]
```diff
@@ -42,6 +42,9 @@ uuid = { version = "0.8", features = ["v4"], optional = true }
...
+encode_unicode = { version = "0.3.6", optional = true }
```
Comment (StephenWakely) [inline lib/vrl/stdlib/Cargo.toml:46]: This isn't used either...

**Review by JeanMertz — APPROVED** [review]
Nice work 👍

**Review by jszwedko — APPROVED** [review]
I think those two benchmarks are noisy since we've seen them on other PRs too.

cc/ @blt we should probably look at those to avoid people going blind to the benchmarks. Or maybe just bump the sensitivity for now?

**Comment by blt** [comment]
> I think those two benchmarks are noisy since we've seen them on other PRs too.

I'll add an erratic filter. The soaks aren't noisy so much as vector's behavior is erratic enough to make samples distinct in a way that is more than down to randomness, which is to say, vector doesn't function reliably in these configurations. It's an important signal, one I didn't really anticipate before we started looking at this kind of thing. But, uh, not actionable in these PRs either. 

> cc/ @blt we should probably look at those to avoid people going blind to the benchmarks. Or maybe just bump the sensitivity for now?

Our current thresholds are as insensitive as I'd like to see us go.

**Pull request merged**